Automattic/woocommerce-subscriptions-core
# [HPOS] Allow admins to filter the Subscriptions List Table when HPOS is enabled.

Fixes #334 

## Description






The Subscriptions List Table allows store managers to filter subscriptions by product, customer and payment method, however, these filters weren't working when HPOS is enabled:

![image](https://user-images.githubusercontent.com/2275145/211227361-068bfa7e-2eb5-4623-8e90-e403f3bc6479.png)

On non-HPOS stores, we are hooking onto the `request` filter and modifying the WP Query to return results based on the filters set in GET params. However this `request` filter isn't used on the new WC Orders List Table page.

On stores that have HPOS enabled, WooCommerce has provided a new hook that lets us filter the table results by modifying the params sent to `wc_get_orders()`.

This PR fixes being able to filter the Subscriptions List Table by implementing the following changes:
- New function to the `woocommerce_shop_subscription_list_table_prepare_items_query_args` hook which filters the query args that get passed to `wc_get_orders()`.
- Moved the code that was used to filter the `request` to filter customers, products and payment methods into separate functions.
- These functions can be repurposed to work for both filtering the List Table for stores that have HPOS enabled and disabled.
- Updated the DB queries called in `wcs_get_subscriptions_for_product()` to also work on stores that have HPOS enabled.

## How to test this PR





1. Purchase multiple different subscription products to help with testing the filtering by product
2. Manually create/add a subscription via the Admin page with a manual payment method
3. Go to **WooCommerce > Subscriptions**
4. With HPOS enabled:
5. Filter the subscriptions table by product
6. Filter the subscriptions table by customer (also check for customers that have no subscriptions)
7. Filter the subscriptions table by payment method (check "none", "woocommerce payments" and "manual payments")
8. Run through the same tests with HPOS disabled.

## Product impact
